kv/dist_sender.go

```diff
@@ -480,6 +472,131 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
 	return err
 }
 
+// sendAttempt is invoked by Send and handles retry logic and cache eviction
+// for a call sent to a single range.
+func (ds *DistSender) sendAttempt(desc *proto.RangeDescriptor, call client.Call) (retry.Status, error) {
+	leader := ds.leaderCache.Lookup(proto.RaftID(desc.RaftID))
+
+	// Try to send the call.
+	replicas := newReplicaSlice(ds.gossip, desc)
+
+	// Rearrange the replicas so that those replicas with long common
+	// prefix of attributes end up first. If there's no prefix, this is a
+	// no-op.
+	order := ds.optimizeReplicaOrder(replicas)
+
+	args := call.Args
+	reply := call.Reply
+
+	// If this request needs to go to a leader and we know who that is, move
+	// it to the front.
+	if !(proto.IsRead(args) && args.Header().ReadConsistency == proto.INCONSISTENT) &&
+		leader.StoreID > 0 {
+		if i := replicas.FindReplica(leader.StoreID); i >= 0 {
+			replicas.MoveToFront(i)
+			order = rpc.OrderStable
+		}
+	}
+
+	rpcErr := ds.sendRPC(desc.RaftID, replicas, order, args, reply)
+
+	// For an RPC error to occur, we must've been unable to contact any
+	// replicas. In this case, likely all nodes are down (or not getting back
+	// to us within a reasonable amount of time).
+	// We may simply not be trying to talk to the up-to-date replicas, so
+	// clearing the descriptor here should be a good idea.
+	// TODO(tschottdorf): If a replica group goes dead, this will cause clients
+	// to put high read pressure on the first range, so there should be some
+	// rate limiting here.
+	if rpcErr != nil {
+		ds.rangeCache.EvictCachedRangeDescriptor(args.Header().Key, desc)
+	}
+
+	if err := util.FirstError(rpcErr, reply.Header().GoError()); err != nil {
```

Comment:
there was a discussion here that github decided to kill:
> @tamird: why do we need to do type assertions on rpcErr if the only failure case is explained and handled immediately above?
> @tschottdorf: I don't see any type assertions. The rpcErr is the error returned by the RPC call, and we don't want to mix it up with the error contained in a successful RPC call's response.
> @tamird: right, but your call to util.FirstError(rpcErr,...) might return rpcErr, so that err == rpcErr, and then you do type assertions below.